Source/DafnyCore/AST/Types/UserDefinedType.cs

```diff
@@ -0,0 +1,509 @@
+        // The type's definition is hidden from the current scope, so we
+        // have to assume the type may involve references.
+        return true;
+      } else if (TypeArgs.Any(ta => ta.ComputeMayInvolveReferences(visitedDatatypes))) {
+        return true;
+      } else if (visitedDatatypes != null && visitedDatatypes.Contains(dt)) {
+        // we're in the middle of looking through the types involved in dt's definition
+        return false;
+      } else {
+        visitedDatatypes ??= new HashSet<DatatypeDecl>();
+        visitedDatatypes.Add(dt);
+        return dt.Ctors.Any(ctor => ctor.Formals.Any(f => f.Type.ComputeMayInvolveReferences(visitedDatatypes)));
+      }
+    } else if (ResolvedClass is TypeSynonymDeclBase) {
+      var t = (TypeSynonymDeclBase)ResolvedClass;
+      if (t.Characteristics.ContainsNoReferenceTypes) {
+        // There's an explicit "(!new)" annotation on the type.
+        return false;
+      } else if (t.IsRevealedInScope(Type.GetScope())) {
+        // The type's definition is available in the scope, so consult the RHS type
+        return t.RhsWithArgument(TypeArgs).ComputeMayInvolveReferences(visitedDatatypes);
+      } else {
+        // The type's definition is hidden from the current scope and there's no explicit "(!new)", so we
+        // have to assume the type may involve references.
+        return true;
+      }
+    } else if (ResolvedClass is TypeParameter typeParameter) {
+      if (visitedDatatypes != null) {
+        // Datatypes look at the type arguments passed in, so we ignore their formal type parameters.
+        // See comment above and in Type.ComputeMayInvolveReferences.
+        Contract.Assert(typeParameter.Parent is DatatypeDecl);
+        return false;
+      } else {
+        return !typeParameter.Characteristics.ContainsNoReferenceTypes;
+      }
+    } else if (ResolvedClass is AbstractTypeDecl opaqueTypeDecl) {
+      return !opaqueTypeDecl.Characteristics.ContainsNoReferenceTypes;
+    }
+    Contract.Assume(false);  // unexpected or not successfully resolved Type
+    return true;
+  }
+
+  public override List<Type> ParentTypes() {
+    return ResolvedClass != null ? ResolvedClass.ParentTypes(TypeArgs) : base.ParentTypes();
+  }
+
+  public override bool IsSubtypeOf(Type super, bool ignoreTypeArguments, bool ignoreNullity) {
+    super = super.NormalizeExpandKeepConstraints();
+
+    // Specifically handle object as the implicit supertype of classes and traits.
+    // "object?" is handled by Builtins rather than the Type hierarchy, so unfortunately
+    // it can't be returned in ParentTypes().
+    if (super.IsObjectQ) {
+      return IsRefType;
+    } else if (super.IsObject) {
+      return ignoreNullity ? IsRefType : IsNonNullRefType;
+    }
+
+    return base.IsSubtypeOf(super, ignoreTypeArguments, ignoreNullity);
+  }
+
+  public IToken NameToken => tok;
+  public IEnumerable<IDeclarationOrUsage> GetResolvedDeclarations() {
```

Comment:
The addition of this method and extending the related interface `IHasUsages` is the only modification made to this class.